File: TypeScript/374.猜数字大小.ts
function guess(pick: number): number { return pick; };

/** 
 * Forward declaration of guess API.
 * @param {number} num   your guess
 * @return 	            -1 if num is lower than the guess number
 *			             1 if num is higher than the guess number
 *                       otherwise return 0
 * var guess = function(num) {}
 */


function guessNumber(n: number): number {
    let [left, right] = [1, n];
    while (left <= right) {
        const mid = Math.floor((right - left) / 2 + left);
        const g = guess(mid);
        if (g == 0)
            return mid;
        else if (g == -1) {  // num < pick ; pick > num
            right = mid - 1;
        } else if (g == 1) {
            left = mid + 1;
        }
    }
    return -1;
};

File: 374.猜数字大小.cc
int guess(int pick);

/** 
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return 	     -1 if num is lower than the guess number
 *			      1 if num is higher than the guess number
 *               otherwise return 0
 * int guess(int num);
 */

class Solution {
public:
    int guessNumber(int n) {
        int left = 1, right = n;
        while (left <= right) {
            int mid = (right - left) / 2 + left;
            int g = guess(mid);
            if (g == 0) {
                return mid;
            } else if (g == -1) {
                right = mid - 1;
            } else if (g == 1) {
                left = mid + 1;
            }
        }
        return -1;
    }
};
